test/concurrent.cc

```diff
@@ -0,0 +1,717 @@
+#include "stout/concurrent.h"
+
+#include <deque>
+
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+#include "stout/callback.h"
+#include "stout/collect.h"
+#include "stout/eventual.h"
+#include "stout/range.h"
+#include "stout/reduce.h"
+#include "stout/stream-for-each.h"
+#include "stout/terminal.h"
+#include "stout/then.h"
+
+namespace eventuals = stout::eventuals;
+
+using stout::Callback;
+
+using stout::eventuals::Collect;
+using stout::eventuals::Concurrent;
+using stout::eventuals::Eventual;
+using stout::eventuals::Interrupt;
+using stout::eventuals::Iterate;
+using stout::eventuals::Loop;
+using stout::eventuals::Map;
+using stout::eventuals::Range;
+using stout::eventuals::Reduce;
+using stout::eventuals::Stream;
+using stout::eventuals::StreamForEach;
+using stout::eventuals::Terminate;
+using stout::eventuals::Then;
+
+using testing::Contains;
+using testing::UnorderedElementsAre;
+
+TEST(ConcurrentTest, Success) {
+  std::deque<Callback<>> callbacks;
+
+  auto e = [&]() {
+    return Iterate({1, 2})
+        | Concurrent([&]() {
+             struct Data {
+               void* k;
+               int i;
+             };
+             return Map(Eventual<std::string>(
+                 [&, data = Data()](auto& k, int i) mutable {
+                   using K = std::decay_t<decltype(k)>;
+                   data.k = &k;
+                   data.i = i;
+                   callbacks.emplace_back([&data]() {
+                     static_cast<K*>(data.k)->Start(std::to_string(data.i));
+                   });
+                 }));
+           })
+        | Collect<std::vector<std::string>>();
+  };
+
+  auto [future, k] = Terminate(e());
+
+  k.Start();
+
+  ASSERT_EQ(2, callbacks.size());
+
+  EXPECT_EQ(
+      std::future_status::timeout,
+      future.wait_for(std::chrono::seconds(0)));
```

Comment:
I really like how you check for the current eventual status here and in `ASSERT_EQ(2, callbacks.size());`! That's very smart. I would appreciate the comments though!